gpt4all-chat/database.cpp

```diff
@@ -1038,6 +1038,53 @@ void Database::handleDocumentError(const QString &errorMessage,
     qWarning() << errorMessage << document_id << document_path << error;
 }
 
+static QString recursiveTextSplitter(QString &text, int &splitLevel, size_t maxChars,
+    const QList<QString> &splitters, int splitterIndex = 0)
+{
+    // If our text is below the limit, then return it
+    if (text.size() <= maxChars) {
+        splitLevel = splitterIndex;
+        const QString chunk = text;
+        text.clear();
+        return chunk;
+    }
+
+    // If we can't split further, then slice and return what we can
+    if (splitterIndex >= splitters.size()) {
+        splitLevel = -1;
+        const QString chunk = text.first(maxChars);
+        text.remove(0, chunk.size());
+        return chunk;
+    }
+
+    // Split the string with current splitter
+    const QString splitter = splitters.at(splitterIndex);
+    const QList<QString> parts = text.split(splitter);
+    Q_ASSERT(parts.size() > 0);
+
+    // If the first part is already above our limit, then move on to the next splitter
+    if (parts.first().size() > maxChars)
+        return recursiveTextSplitter(text, splitLevel, maxChars, splitters, splitterIndex + 1);
+
+    // Otherwise, we have our split level now accumulate as much as we can on this level
+    QList<QString>::const_iterator it = parts.constBegin();
+    QString chunk = *it;
+    for (++it; it != parts.constEnd(); ++it) {
```

Comment:
Abuse of `for` syntax. If you really want to start at `it + 1`, use `chunk = *it++` above, or put `++it` on its own line. Couldn't this be a range-for ("for each"), anyway?

Reply:
Yeah, again I'm refactoring to get rid of the split so this will change. Thanks for feedback though.